c/include/libsbp/cpp/message_handler.h

```diff
@@ -0,0 +1,146 @@
+}
+
+namespace details {
+
+/**
+ * Recursive interface type for defining the interface functions for `MessageHandler`.
+ *
+ * These types define a virtual `operator()` for handling a specific SBP message type,
+ * as well as a function for registering it as a SBP callback.
+ *
+ * @note These types aren't meant to be used directly by application code.
+ *
+ * @tparam MsgType The type of SBP message to define an interface for
+ * @tparam OtherTypes Other types to recursively define interfaces for
+ */
+template<typename MsgType, typename... OtherTypes>
+class CallbackInterface : CallbackInterface<OtherTypes...> {
+ public:
+  CallbackInterface() = default;
+
+  using CallbackInterface<OtherTypes...>::handle_sbp_msg;
+  virtual void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const MsgType& msg) = 0;
+
+ protected:
+  void register_callback(sbp_state_t *state, sbp_msg_callbacks_node_t nodes[]) {
+    sbp_register_callback(state,
+        sbp::MessageTraits<MsgType>::id,
+        &sbp_cb_passthrough<MsgType, CallbackInterface, &CallbackInterface::handle_sbp_msg>,
+        this,
+        &nodes[0]);
+    CallbackInterface<OtherTypes...>::register_callback(state, &nodes[1]);
+  }
+};
+
+template<typename MsgType>
+class CallbackInterface<MsgType> {
+ public:
+  CallbackInterface() = default;
+
+  virtual void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const MsgType& msg) = 0;
+ protected:
+  void register_callback(sbp_state_t *state, sbp_msg_callbacks_node_t nodes[]) {
+    sbp_register_callback(state,
+                          sbp::MessageTraits<MsgType>::id,
+                          &sbp_cb_passthrough<MsgType, CallbackInterface, &CallbackInterface::handle_sbp_msg>,
+                          this,
+                          &nodes[0]);
+  }
+};
+
+} // namespace details
+
+/**
+ * Base type for defining classes that handle SBP messages
+ *
+ * Application classes should derive from this class if they wish to handle
+ * SBP messages with a member function. `MessageHandler` instantiates all of
+ * the callback nodes, and registers the member functions with the given
+ * `sbp_state_t`.
+ *
+ * Classes that derive from `MessageHandler` need to implement
+ *   void handle_sbp_msg(uint16_t sender_id, uint8_t message_length, const MsgType& msg) const;
+ * for each `MsgType` in the list of message types given as template parameters.
+ *
+ * @note It should not matter if the class derives publicly or privately from `MessageHandler`
+ *
+ * @tparam MsgTypes List of SBP message types to register callbacks for
+ */
+template<typename... MsgTypes>
+ class MessageHandler : public details::CallbackInterface<MsgTypes...> {
+    static constexpr size_t kMsgCount = sizeof...(MsgTypes);
+
+    State &state_;
+    std::array<sbp_msg_callbacks_node_t, kMsgCount> callback_nodes_;
+
+  public:
+
+    explicit MessageHandler(State *state) : details::CallbackInterface<MsgTypes...>(), state_(*state), callback_nodes_() {
```

Comment:
Is there a reason the ctor argument isn't `State&`?

Reply:
The `State` needs to be mutable, and our C++ guidelines say any non-`const` parameter should be passed as a pointer.